src/bin/css.rs

```diff
@@ -0,0 +1,122 @@
+
+extern crate gdk;
+extern crate gio;
+extern crate glib;
+extern crate gtk;
+
+use gio::prelude::*;
+use gtk::prelude::*;
+
+use std::env::args;
+
+// Basic CSS: we change background color, we set font color to black and we set it as bold.
+const STYLE: &'static str = "
+#entry1 {
+    background-image: -gtk-gradient (linear,
+                                     0 0, 1 0,
+                                     color-stop(0, #f00),
+                                     color-stop(1, #0f0));
+    color: blue;
+    font-weight: bold;
+}
+
+button {
+    /* If we don't put it, the yellow background won't be visible */
+    background-image: none;
+}
+#label1:hover {
+    transition: 500ms;
+    color: red;
+    background-color: yellow;
+}
+
+combobox button.combo box {
+    padding: 5px;
+}
+combobox box arrow {
+    -gtk-icon-source: none;
+    border-left: 5px solid transparent;
+    border-right: 5px solid transparent;
+    border-top: 5px solid black;
+}";
+
+// upgrade weak reference or return
+#[macro_export]
+macro_rules! upgrade_weak {
+    ($x:ident, $r:expr) => {{
+        match $x.upgrade() {
+            Some(o) => o,
+            None => return $r,
+        }
+    }};
+    ($x:ident) => {
+        upgrade_weak!($x, ())
+    };
+}
+
+fn build_ui(application: &gtk::Application) {
+    let window = gtk::ApplicationWindow::new(application);
+
+    window.set_title("CSS");
+    window.set_position(gtk::WindowPosition::Center);
+
+    window.connect_delete_event(|win, _| {
+        win.destroy();
+        Inhibit(false)
+    });
+
+    // The container container.
+    let vbox = gtk::Box::new(gtk::Orientation::Vertical, 0);
+
+    let label = gtk::Button::new_with_label("hover me!");
+    // We need to name it in order to apply CSS on it.
+    gtk::WidgetExt::set_name(&label, "label1");
+
+    let entry = gtk::Entry::new();
+    // We need to name it in order to apply CSS on it.
+    gtk::WidgetExt::set_name(&entry, "entry1");
+    entry.set_text("Some text");
+
+    let combo = gtk::ComboBoxText::new();
+    combo.append_text("option 1");
+    combo.append_text("option 2");
+    combo.append_text("option 3");
+    combo.set_active(0);
+
+    vbox.add(&label);
+    vbox.add(&entry);
+    vbox.add(&combo);
+    // Then we add the container inside our window.
+    window.add(&vbox);
+
+    application.connect_activate(move |_| {
+        window.show_all();
+    });
+}
+
+fn main() {
+    let application = gtk::Application::new("com.github.css",
+                                            gio::ApplicationFlags::empty())
+                                       .expect("Initialization failed...");
+
+    application.connect_startup(|app| {
+        // The CSS "magic" happens here.
+        let provider = gtk::CssProvider::new();
+        provider
+            .load_from_data(STYLE.as_bytes())
+            .expect("Failed to load CSS");
+        gtk::StyleContext::add_provider_for_screen(
```

Comment:
Explain what the meaning of the CSS providers and the style context is, how it works and what these priorities are.